### PR TITLE
Prevent potential OOB read in EC_GROUP_get_basis_type

### PR DESCRIPTION
**Pull request by guidovranken**
##### Description of change

When checking the elements of a fixed-size array (EC_GROUP's int poly[6]), do not continue iterating after its last element has been reached; instead, check up to the last element, and break out of the loop once this position is reached. This will prevent out-of-bounds access in case each element of the poly[] array is non-zero.

### CODE REVIEW

File: crypto/ec/ec_asn1.c [hunk 1]
```diff
@@ -15,15 +15,17 @@
...
     if (EC_METHOD_get_field_type(EC_GROUP_method_of(group)) !=
         NID_X9_62_characteristic_two_field)
         /* everything else is currently not supported */
         return 0;
 
-    while (group->poly[i] != 0)
-        i++;
+    /* Find the last non-zero element of group->poly[] */
+    for (i = 0;
+        i < sizeof(group->poly)/sizeof(group->poly[0]) && group->poly[i] != 0;
```
Comment (richsalz) [inline crypto/ec/ec_asn1.c]: ```OSSL_NELEM```

Comment (guidovranken) [inline crypto/ec/ec_asn1.c]: Right.

File: crypto/ec/ec_asn1.c [hunk 2]
```diff
@@ -15,15 +15,17 @@
...
         NID_X9_62_characteristic_two_field)
         /* everything else is currently not supported */
         return 0;
 
-    while (group->poly[i] != 0)
-        i++;
+    /* Find the last non-zero element of group->poly[] */
+    for (i = 0;
+        i < OSSL_NELEM(group->poly) & group->poly[i] != 0;
+        i++) ;
```
Comment (levitte) [inline crypto/ec/ec_asn1.c]: Nit pick.  The second and third line should indent by one more space, so they're all "inside" the parentheses.

Comment (richsalz) [inline crypto/ec/ec_asn1.c]: and also, use an explicit continue (on a line by itself)

**Comment by richsalz** [comment]
+1 from me.  you good with this, @levitte ?

**Comment by richsalz** [comment]
1.0.2 ea63fdd, 1.1.0 b757d2d, master 57f48f9.

thank you!

File: crypto/ec/ec_asn1.c [hunk 3]
```diff
@@ -15,15 +15,18 @@
...
     if (EC_METHOD_get_field_type(EC_GROUP_method_of(group)) !=
         NID_X9_62_characteristic_two_field)
         /* everything else is currently not supported */
         return 0;
 
-    while (group->poly[i] != 0)
-        i++;
+    /* Find the last non-zero element of group->poly[] */
+    for (i = 0;
+         i < OSSL_NELEM(group->poly) & group->poly[i] != 0;
```
Comment (levitte) [inline crypto/ec/ec_asn1.c]: Damnit, didn't see this before...  Surely, the lonely `&` is supposed to have some company, i.e. `&&`

Comment (guidovranken) [inline crypto/ec/ec_asn1.c]: Woops. I should have tested it. Do I create a new PR now?

Comment (richsalz) [inline crypto/ec/ec_asn1.c]: +1 for @levitte to fix it.
or if he gives me the +1 I'll do it for all branches.

Comment (levitte) [inline crypto/ec/ec_asn1.c]: Done.

master: 50799f3558981eac0482d3ea77b21c58b56d4871
1.1.0: c1d0d2e690d2a33216de6c5920c81d9852080d36
1.0.2: 6161c5d81b66f38a8a65b78379182a42a9819cc6

**Pull request closed without merging**